### PR TITLE
WIP: Replacing flexc++ with re2c

### PR DESCRIPTION
**Pull request by certik**
TODO:

* [x] Make the PR nice and clean
* [ ] Squash the commits into a few nice commits

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -35,10 +35,10 @@ endif()
 # Enable C++11 support in all compilers. SymEngine will not compile unless
 # the C++11 support is enabled.
 if (CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang|Intel")
-    set(CXX11_OPTIONS "-std=c++11")
+    set(CXX11_OPTIONS "-std=c++14")
```
Comment (isuruf) [inline CMakeLists.txt]: Is this necessary?

Comment (certik) [reply CMakeLists.txt]: No. I only needed it for `make_unique`, which I will find some workaround for.

**Comment by certik** [comment]
re2c is only a lexer. The parser stays the same. I am going to play with the parser later after this is done. So it's very impressive that a better lexer can deliver so much speedup.

File: CMakeLists.txt (hunk 1 above)
Comment (certik) [reply CMakeLists.txt]: Fixed, I switched back to C++11.

**Comment by certik** [comment]
I was wondering about the `&&yy2` syntax (`yy2` is a label). gcc and clang seem to support it, but MSVC does not. So I converted the tables to if statements. One can also just use switch statements. The differences were minor in terms how gcc can optimize the code. It converts switch statements to jump tables anyway.

**Comment by certik** [comment]
@isuruf this is ready for review. Where should we move the `make_unique` implementation in `unique.h`? 

The tokenizer is now thread-safe. I polished it up, I am pretty happy with it now.

File: symengine/parser/tokenizer.re [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+            ident = char (char | dig)*;
+            numeric = (dig*"."?dig+([eE][-+]?dig+)?) | (dig+".");
+            implicitmul = numeric ident;
+
+            *   { printf("ERR unknown token\n"); return 0; }
+            end {
+                    if (in.lim - in.tok == YYMAXFILL) {
+                        return 0;
+                    } else {
+                        printf("ERR NULL.\n");
```
Comment (certik) [inline symengine/parser/tokenizer.re]: These two print statements are the only ugly parts of the code. Essentially they signal an error. I think we should raise an exception instead?

Comment (isuruf) [reply symengine/parser/tokenizer.re]: Yes, a `SymEngine::ParseError`.

**Comment by isuruf** [comment]
We could move to https://github.com/symengine/symengine/blob/master/symengine/dict.h where utility functions are. (Maybe we should rename that file)

**Comment by certik** [comment]
The other consideration is that currently the tokenizer has a buffer, and is able to read from an input stream of arbitrary length (doesn't have to fit into memory), and it correctly buffers it, and reloads the buffer, etc.

However, if the input is so long that the text file does not fit into memory, then even if tokenizer can process it, I think the parser will run out of memory. So I think for our use case, we can assume that things fit into memory.

In that case, I should further simplify the tokenizer. This will remove I think two error conditions that currently can occur, but if there is no buffer, then I think the error conditions could not occur. So things become simpler and more robust. If we can fit into memory. Which I think we can.

@isuruf let me know what you think.

File: symengine/parser/tokenizer.re (hunk 2 above)
Comment (certik) [reply symengine/parser/tokenizer.re]: My understanding that both of these error conditions will disappear after I rework the tokenizer per my comment https://github.com/symengine/symengine/pull/1586#issuecomment-517745895.

**Comment by isuruf** [comment]
It's okay to assume that strings fit into memory. If it doesn't, the symengine Basic object will not fit into memory either.

**Comment by certik** [comment]
I agree. So let me rework this. It will greatly simplify things, I just need to carefully test the new implementation.

**Comment by certik** [comment]
As commented [here](https://github.com/symengine/symengine/pull/1584#issuecomment-517819809), this is now as fast or faster than flex. That's due to the simplifications coming from end of stream checking.

**Comment by certik** [comment]
@isuruf would you mind reviewing it one more time? Things greatly simplified and got even a bit faster.

If this looks good to you, I will squash the commits.

File: bin/travis_clang_format.sh [hunk 3]
```diff
@@ -13,7 +13,7 @@ if [[ "${TEST_CLANG_FORMAT}" == "yes" ]]; then
         exit 1
     fi
 
-    FILES=`git ls-files | grep -E "\.(cpp|h|hpp|c)$" | grep -Ev "symengine/utilities" | grep -Ev "cmake/"`
+    FILES=`git ls-files | grep -E "\.(cpp|h|hpp|c)$" | grep -Ev "symengine/utilities" | grep -Ev "cmake/" | grep -Ev "symengine/parser/tokenizer.cpp"`
```
Comment (isuruf) [inline bin/travis_clang_format.sh]: Can you revert this change as `bin/test_format_local.sh` is updated to format this file as well?

Comment (isuruf) [reply bin/travis_clang_format.sh]: We can format that file.

File: symengine/CMakeLists.txt [hunk 4]
```diff
@@ -7,17 +7,18 @@ if (WITH_SYMENGINE_TEUCHOS)
...
                       DEPENDS parser/parser.yy parser/parser.ih
                       WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/parser
                      )
 
-    add_custom_command(OUTPUT parser/scanner.cpp
-                      COMMAND ${FLEXCPP_EXECUTABLE} --namespace=SymEngine scanner.ll
-                      DEPENDS parser/scanner.ll parser/scanner.h
-                      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/parser/
+    add_custom_command(OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/parser/tokenizer.cpp
+                      COMMAND ${RE2C_EXECUTABLE} -W -b tokenizer.re -o tokenizer.cpp
```
Comment (isuruf) [inline symengine/CMakeLists.txt]:
```suggestion
                      COMMAND ${RE2C_EXECUTABLE} -W  --no-generation-date -b tokenizer.re -o tokenizer.cpp
```

File: bin/travis_clang_format.sh (hunk 3 above)
Comment (certik) [reply bin/travis_clang_format.sh]: OK. I formatted the file and reverted this.

File: symengine/parser/tokenizer.cpp [hunk 5]
```diff
@@ -0,0 +1,380 @@
+/* Generated by re2c 1.1.1 */
```
Comment (isuruf) [inline symengine/parser/tokenizer.cpp]: I ran the parser and formatted and got the exact same file now with re2c 1.1.1.

Comment (certik) [reply symengine/parser/tokenizer.cpp]: Thanks.

**Comment by certik** [comment]
Closing in favor of #1588.
